src/libraries/System.Private.CoreLib/src/System/IO/Strategies/FileStreamValueTaskSource.cs

```diff
@@ -0,0 +1,252 @@
+            {
+                // Extract the AwaitableProvider from the overlapped.  The state in the overlapped
+                // will either be a AsyncWindowsFileStreamStrategy (in the case where the preallocated overlapped was used),
+                // in which case the operation being completed is its _currentOverlappedOwner, or it'll
+                // be directly the AwaitableProvider that's completing (in the case where the preallocated
+                // overlapped was already in use by another operation).
+                object? state = ThreadPoolBoundHandle.GetNativeOverlappedState(pOverlapped);
+                Debug.Assert(state is (AsyncWindowsFileStreamStrategy or FileStreamValueTaskSource));
+                FileStreamValueTaskSource valueTaskSource = state switch
+                {
+                    AsyncWindowsFileStreamStrategy strategy => strategy._currentOverlappedOwner!, // must be owned
+                    _ => (FileStreamValueTaskSource)state
+                };
+                Debug.Assert(valueTaskSource != null);
+                Debug.Assert(valueTaskSource._overlapped == pOverlapped, "Overlaps don't match");
+
+                // Handle reading from & writing to closed pipes.  While I'm not sure
+                // this is entirely necessary anymore, maybe it's possible for
+                // an async read on a pipe to be issued and then the pipe is closed,
+                // returning this error.  This may very well be necessary.
+                ulong packedResult;
+                if (errorCode != 0 && errorCode != Interop.Errors.ERROR_BROKEN_PIPE && errorCode != Interop.Errors.ERROR_NO_DATA)
+                {
+                    packedResult = ((ulong)FileStreamHelpers.ResultError | errorCode);
+                }
+                else
+                {
+                    packedResult = ((ulong)FileStreamHelpers.ResultSuccess | numBytes);
+                }
+
+                // Stow the result so that other threads can observe it
+                // And, if no other thread is registering cancellation, continue
+                if (Interlocked.Exchange(ref valueTaskSource._result, (long)packedResult) == FileStreamHelpers.NoResult)
+                {
+                    // Successfully set the state, attempt to take back the callback
+                    if (Interlocked.Exchange(ref valueTaskSource._result, FileStreamHelpers.CompletedCallback) != FileStreamHelpers.NoResult)
+                    {
+                        // Successfully got the callback, finish the callback
+                        valueTaskSource.CompleteCallback(packedResult);
+                    }
+                    // else: Some other thread stole the result, so now it is responsible to finish the callback
+                }
+                // else: Some other thread is registering a cancellation, so it *must* finish the callback
```

Comment:
All this logic looks more complicated than it needs to be.  I realize you're just following the same flow as what was already there, so nothing needs to be changed in this PR.  But we should revisit all of this synchronization later; I suspect we can do it more succinctly and efficiently.

Reply:
I opened an issue to follow up on this and on the caching of the IValueTaskSource.
https://github.com/dotnet/runtime/issues/50972